taglib/toolkit/tvariant.cpp

```diff
@@ -0,0 +1,391 @@
+
+using namespace TagLib;
+
+namespace {
+
+// The number and order of the template parameters must correspond to the
+// enum values in Variant::Type!
+using StdVariantType = std::variant<
+  std::monostate,
+  bool,
+  int,
+  unsigned int,
+  long long,
+  unsigned long long,
+  double,
+  TagLib::String,
+  TagLib::StringList,
+  TagLib::ByteVector,
+  TagLib::ByteVectorList,
+  List<TagLib::Variant>,
+  Map<TagLib::String, TagLib::Variant>
+>;
+
+template<typename T>
+T getVariantValue(StdVariantType *v, bool *ok)
+{
+  if(const auto valPtr = std::get_if<T>(v)) {
+    if(ok) {
+      *ok = true;
+    }
+    return *valPtr;
+  }
+  if(ok) {
+    *ok = false;
+  }
+  return {};
+}
+
+// Visitor to print a possibly recursive Variant to an ostream.
+// The representation is JSON with hex strings for ByteVector.
+class OStreamVisitor {
+public:
+  OStreamVisitor(std::ostream &os) : s(os)
+  {
+  }
+
+  void operator()(std::monostate v)
+  {
+    s << "null";
+  }
+
+  void operator()(bool v)
+  {
+    s << (v ? "true" : "false");
+  }
+
+  void operator()(int v)
+  {
+    s << v;
+  }
+
+  void operator()(unsigned int v)
+  {
+    s << v;
+  }
+
+  void operator()(long long v)
+  {
+    s << v;
+  }
+
+  void operator()(unsigned long long v)
+  {
+    s << v;
+  }
+
+  void operator()(double v)
+  {
+    s << v;
+  }
+
+  void operator()(const TagLib::String &v)
+  {
+    s << '"';
+    for (char c : v.to8Bit()) {
+      if(c == '"') {
+        s << "\\\"";
+      }
+      else {
+        s << c;
+      }
+    }
+    s << '"';
+  }
+
+  void operator()(const TagLib::StringList &v)
+  {
+    s << '[';
+    for(auto it = v.cbegin(); it != v.cend(); ++it) {
+      if(it != v.cbegin()) {
+        s << ", ";
+      }
+      operator()(*it);
+    }
+    s << ']';
+  }
+
+  void operator()(const TagLib::ByteVector &v)
+  {
+    s << '"';
+    for(char c : v) {
+      s << "\\x" << std::setfill('0') << std::setw(2) << std::right << std::hex
+        << (static_cast<int>(c) & 0xff);
+    }
+    s << std::dec << '"';
+  }
+
+  void operator()(const TagLib::ByteVectorList &v)
+  {
+    s << '[';
+    for(auto it = v.cbegin(); it != v.cend(); ++it) {
+      if(it != v.cbegin()) {
+        s << ", ";
+      }
+      operator()(*it);
+    }
+    s << ']';
+  }
+
+  void operator()(const List<TagLib::Variant> &v) {
+    s << '[';
+    for(auto it = v.cbegin(); it != v.cend(); ++it) {
+      if(it != v.cbegin()) {
+        s << ", ";
+      }
+      s << *it;
+    }
+    s << ']';
+  }
+
+  void operator()(const Map<TagLib::String, TagLib::Variant> &v)
+  {
+    s << '{';
+    for(auto it = v.cbegin(); it != v.cend(); ++it) {
+      if(it != v.cbegin()) {
+        s << ", ";
+      }
+      operator()(it->first);
+      s << ": ";
+      s << it->second;
+    }
+    s << '}';
+  }
+
+private:
+  std::ostream &s;
+};
+
+} // namespace
+
+class Variant::VariantPrivate
+{
+public:
+  VariantPrivate() = default;
+  VariantPrivate(StdVariantType v) : data(v) {}
```

Comment:
why is this not moved?

Reply:
I actually do not like move semantics in C++. Reconsidering it again, I changed it back to a good old const reference. Since all types contained in the variant are cheap to copy, this is not really a rewarding use of move semantics. I tried it with `std::move()` before and GCC was generating larger code. And I do not want to open the door for more `std::move()` "improvements" in TagLib. After all, TagLib uses implicitly shared strings and containers, which are in my opinion a less error-prone and more user friendly way to avoid costly copies and reallocations.